### PR TITLE
Watch secondary resources of Kafka components

### PR DESCRIPTION
**Pull request by aliok**
Bases https://github.com/openshift-knative/serverless-operator/pull/493

Similar to other PRs in the umbrella ticket, https://github.com/openshift-knative/serverless-operator/issues/486, I didn't want to wait until the CI is green and that's merged. This is a WIP PR to see in advance if things fail and if possibly get some reviews.

-------------------------
## Changes
This PR creates watches for the secondary resources that are created.

Also, I noticed there were a lot of duplicate code, so I created reusable functions following in a separate PR (https://github.com/openshift-knative/serverless-operator/pull/513):
- injecting owner annotations
- creating reconciliation enqueue request for watched resources
- finding the resource types to watch
-------------------------

As written in the umbrella ticket,  https://github.com/openshift-knative/serverless-operator/issues/486, there's still a lot to do.

Verification:

1. Install eventing

2. Enable the controller by uncommenting the stuff in `knative-operator/pkg/controller/add_knativekafka.go`

3. Need to manually create the CRD since it is not in the CSV yet
```
kubectl apply -f knative-operator/deploy/crds/operator_v1alpha1_knativekafka_crd.yaml
```

4. Install:
```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.serverless.openshift.io/v1alpha1
kind: KnativeKafka
metadata:
  name: example
  namespace: knative-eventing
spec:
  source:
    enabled: true
  channel:
    enabled: true
    bootstrapServers: foo.example.com:1234
...
EOS
```

5. Delete any resource created by the operator for Kafka and see if it is recreated.

6. Clean up:
```
kubectl delete knativekafkas example -n knative-eventing
```

### CODE REVIEW

**Comment by matzew** [comment]
@aliok I think this needs a rebase ;-)

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 1]
```diff
@@ -191,22 +246,32 @@ func installKnativeKafkaChannel(instance *operatorv1alpha1.KnativeKafka, apiclie
...
-	manifest, err := mfc.NewManifest(kafkaChannelManifestPath(), apiClient, mf.UseLogger(log.WithName("mf")))
+	manifest, err := rawKafkaChannelManifest(apiClient)
 	if err != nil {
 		return nil, fmt.Errorf("failed to load KafkaChannel manifest: %w", err)
 	}
 
-	manifest, err = manifest.Transform(mf.InjectOwner(instance))
+	transformers := []mf.Transformer{
+		mf.InjectOwner(instance),
+		common.SetOwnerAnnotations(instance.ObjectMeta, common.KafkaOwnerName, common.KafkaOwnerNamespace),
```
Comment (aliok) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:237]: Impl for `common.SetOwnerAnnitations` exists in a later commit (https://github.com/openshift-knative/serverless-operator/pull/496/commits/359a39b0817b8ccc09c134fa2766efa5c48b447f#diff-052673a614af4880bcfbd6c264d9f495R73). Rebasing messed up stuff.

Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:237]: But it is a simple function. Instead of setting the owner annotation with hardcoded keys, it now sets it using the keys passed as arguments.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 2]
```diff
@@ -57,7 +60,59 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+				NamespacedName: types.NamespacedName{Namespace: ownerNamespace, Name: ownerName},
+			}}
+		}
+		return nil
+	})
+
+	gvkToResource := make(map[schema.GroupVersionKind]runtime.Object)
+
+	// Watch for Knative KafkaChannel resources.
+	kafkaChannelManifest, err := rawKafkaChannelManifest(mgr.GetClient())
```
Comment (aliok) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:70]: Manifest was also needed in this function. But it was needed as untransformed, so I created a new `rawKafkaChannelManifest` function that returns the raw manifest. Other places that need the transformed manifest calls another function, which eventually calls `rawKafkaChannelManifest`

**Comment by aliok** [comment]
This is ready for review @matzew @markusthoemmes 
I tried it locally, works fine.

There were some refactorings in this PR, moved them to a follow up PR: https://github.com/openshift-knative/serverless-operator/pull/513

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 3]
```diff
@@ -57,7 +60,59 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
 	}
 
-	// NOT IMPLEMENTED YET
+	err = c.Watch(&source.Kind{Type: &corev1.ConfigMap{}}, &handler.EnqueueRequestForOwner{
+		OwnerType:    &operatorv1alpha1.KnativeKafka{},
+		IsController: true,
+	})
+	if err != nil {
+		return err
+	}
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:69]: Why is this necessary? Are we creating any configmaps specifically that wouldn't be caught by the generic loop below?

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 4]
```diff
@@ -57,7 +60,59 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+	// Watch for Knative KafkaSource resources.
+	kafkaSourceManifest, err := rawKafkaSourceManifest(mgr.GetClient())
+	if err != nil {
+		return err
+	}
+	kafkaSourceResources := kafkaSourceManifest.Resources()
+
+	for i := range kafkaSourceResources {
+		resource := &kafkaSourceResources[i]
+		gvkToResource[resource.GroupVersionKind()] = resource
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:107]:
```suggestion
		gvkToResource[resource.GroupVersionKind()] = &kafkaSourceResources[i]
```

Same above.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 5]
```diff
@@ -215,19 +280,34 @@ func installKnativeKafkaSource(apiclient client.Client) error {
...
+	manifest, err := rawKafkaSourceManifest(apiclient)
 	if err != nil {
-		return mf.Manifest{}, fmt.Errorf("failed to load KafkaSource manifest: %w", err)
+		return nil, fmt.Errorf("failed to load KafkaSource manifest: %w", err)
 	}
-	return manifest, nil
+
+	transformers := []mf.Transformer{
+		setOwnerAnnotations(instance),
+	}
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:303]: Unnecessary intermediary slice.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 6]
```diff
@@ -191,22 +246,32 @@ func installKnativeKafkaChannel(instance *operatorv1alpha1.KnativeKafka, apiclie
...
+}
+
 func kafkaChannelManifest(instance *operatorv1alpha1.KnativeKafka, apiClient client.Client) (*mf.Manifest, error) {
-	manifest, err := mfc.NewManifest(kafkaChannelManifestPath(), apiClient, mf.UseLogger(log.WithName("mf")))
+	manifest, err := rawKafkaChannelManifest(apiClient)
 	if err != nil {
 		return nil, fmt.Errorf("failed to load KafkaChannel manifest: %w", err)
 	}
 
-	manifest, err = manifest.Transform(mf.InjectOwner(instance))
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: Why not keep this format and just keep adding the transformers directly? It takes a variadic argument anyway.

File: knative-operator/pkg/common/constants.go [hunk 7]
```diff
@@ -6,4 +6,6 @@ const (
 	ServingOwnerNamespace            = "serving.knative.openshift.io/ownerNamespace"
 	ServerlessOperatorOwnerName      = "operator.knative.openshift.io/ownerName"
 	ServerlessOperatorOwnerNamespace = "operator.knative.openshift.io/ownerNamespace"
+	KafkaOwnerName                   = "kafka.knative.openshift.io/ownerName"
+	KafkaOwnerNamespace              = "kafka.knative.openshift.io/ownerNamespace"
```
Comment (markusthoemmes) [inline knative-operator/pkg/common/constants.go:10]: Why this group? I'd think this should be `knativekafkas.operator.serverless.openshift.io` to play nice with the group used for the API.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 2 above)
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:87]: We're reading these a few times in the process of the reconciler. Should we maybe create the manifests just once at reconciler initialization and then always use the instance on the reconciler itself to avoid all the duplicated error handling. Plus it'd be a lot quicker.

File: knative-operator/pkg/common/constants.go (hunk 7 above)
Comment (matzew) [reply knative-operator/pkg/common/constants.go:10]: +1

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 3 above)
Comment (matzew) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:69]: The config map for the kafka bootstrap server (backed by the CRD)?

Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:69]: Nah, that will be caught by the generic loop below.

This is a copy paste error

File: knative-operator/pkg/common/constants.go (hunk 7 above)
Comment (aliok) [reply knative-operator/pkg/common/constants.go]: Done

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 6 above)
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: Yep

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 5 above)
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]: Removed

File: knative-operator/pkg/common/constants.go [hunk 8]
```diff
@@ -6,4 +6,6 @@ const (
 	ServingOwnerNamespace            = "serving.knative.openshift.io/ownerNamespace"
 	ServerlessOperatorOwnerName      = "operator.knative.openshift.io/ownerName"
 	ServerlessOperatorOwnerNamespace = "operator.knative.openshift.io/ownerNamespace"
+	KafkaOwnerName                   = "knativekafkas.operator.serverless.openshift.io/ownerName"
+	KafkaOwnerNamespace              = "knativekafkas.operator.serverless.openshift.io/ownerNamespace"
```
Comment (markusthoemmes) [inline knative-operator/pkg/common/constants.go:10]: Thinking about this, this is actually kinda annoying isn't it?

Maybe we should normalize all of them and add a third one that denotes the owner type so can distinguish in the respective operators.

For this specifically that could mean we do these:

```
operator.serverless.openshift.io/ownerName: "foo"
operator.serverless.openshift.io/ownerNamespace: "knative-eventing"
operator.serverless.openshift.io/ownerType: "KnativeKafka"
```

Comment (matzew) [reply knative-operator/pkg/common/constants.go:10]: Do we move all artifacts than to `knative-eventing` ?

Comment (markusthoemmes) [reply knative-operator/pkg/common/constants.go:10]: Nah, that was only an example to make clear what I think this could look like. The namespace would of course be dynamic to the resource.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 2 above)
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]: @markusthoemmes  please have a look at this comment: https://github.com/openshift-knative/serverless-operator/pull/496/commits/41ff1ae8b369f62c883775c8cb51a0d9be135f2c

I did it, but kindaa turned ugly

File: knative-operator/pkg/common/constants.go (hunk 8 above)
Comment (aliok) [reply knative-operator/pkg/common/constants.go:10]: I can do this for KnativeKafka controller only.

For consistency, I would like to change others too, but I am worried a bit about what happens during an upgrade

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 9]
```diff
@@ -241,27 +253,22 @@ func rawKafkaChannelManifest(apiclient client.Client) (mf.Manifest, error) {
...
 }
 
-func kafkaChannelManifest(instance *operatorv1alpha1.KnativeKafka, apiClient client.Client) (*mf.Manifest, error) {
-	manifest, err := rawKafkaChannelManifest(apiClient)
-	if err != nil {
-		return nil, fmt.Errorf("failed to load KafkaChannel manifest: %w", err)
-	}
-
-	manifest, err = manifest.Transform(
+func (r *ReconcileKnativeKafka) kafkaChannelManifest(instance *operatorv1alpha1.KnativeKafka, apiClient client.Client) (*mf.Manifest, error) {
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:256]: The `apiClient` is now unused here (and in the other function probably too)

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 10]
```diff
@@ -354,26 +356,26 @@ func (r *ReconcileKnativeKafka) delete(instance *operatorv1alpha1.KnativeKafka)
...
 			return fmt.Errorf("unable to delete Knative KafkaSource: %w", err)
 		}
 	}
 
 	return nil
 }
 
-func deleteKnativeKafkaChannel(instance *operatorv1alpha1.KnativeKafka, apiclient client.Client) error {
-	manifest, err := kafkaChannelManifest(instance, apiclient)
+func (r *ReconcileKnativeKafka) deleteKnativeKafkaChannel(instance *operatorv1alpha1.KnativeKafka, apiclient client.Client) error {
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:375]: Client will be unused here too I believe.

File: knative-operator/pkg/common/constants.go (hunk 8 above)
Comment (markusthoemmes) [reply knative-operator/pkg/common/constants.go:10]: Yeah we have to do the changes to the others graduall (add them in one release, only start depending on them in the next one).

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 9 above)
Comment (matzew) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:256]: yep - I just checked the `kafkaSourceManifest`, IDEA reports unused

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 11]
```diff
@@ -146,25 +203,25 @@ func (r *ReconcileKnativeKafka) ensureFinalizers(instance *operatorv1alpha1.Knat
 
 // Install Knative Kafka components
 func (r *ReconcileKnativeKafka) installKnativeKafka(instance *operatorv1alpha1.KnativeKafka) error {
-	if err := applyKnativeKafka(instance, r.client); err != nil {
+	if err := r.applyKnativeKafka(instance, r.client); err != nil {
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:206]: I think now the only remaining client usage is in the checkDeployments functions. Should we just pass r.client down in `installKnativeKafkaChannel/Source` and drop the passing through from everywhere?

File: knative-operator/pkg/common/constants.go (hunk 8 above)
Comment (aliok) [reply knative-operator/pkg/common/constants.go:10]: Let me try doing this later ok? Not sure how this would work because we don't have the CR instance at the creation time of the watches

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 9 above)
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]: Fixed

File: knative-operator/pkg/common/constants.go (hunk 8 above)
Comment (markusthoemmes) [reply knative-operator/pkg/common/constants.go:10]: Yes, let's just make sure we do this before we enable this controller. I can send a patch with how I think it should look like too,  you don't have to do everything :).

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 2 above)
Comment (markusthoemmes) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]: How so? I like it!

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 11 above)
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]: Done.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 2 above)
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]: haha ok, everybody has a different taste :D

File: knative-operator/pkg/common/constants.go (hunk 8 above)
Comment (aliok) [reply knative-operator/pkg/common/constants.go:10]: Cool, I will ping you just before I send a PR to enable the controller.

**Comment by aliok** [comment]
Pushed the latest commit that addresses some comments. Please take another look @markusthoemmes  @matzew

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve

Beauitful! :tada:
